odoo/models.py

```diff
@@ -3029,26 +3000,56 @@ def check_access_rule(self, operation):
         if self._uid == SUPERUSER_ID:
             return
 
+        invalid = self - self._filter_access_rule(operation)
+        if not invalid:
+            return
+
+        forbidden = invalid.exists()
+        if forbidden:
+            # the invalid records are (partially) hidden by access rules
+            if self.is_transient():
+                raise AccessError(_('For this kind of document, you may only access records you created yourself.\n\n(Document type: %s)') % (self._description,))
+            else:
+                _logger.info('Access Denied by record rules for operation: %s on record ids: %r, uid: %s, model: %s', operation, forbidden.ids, self._uid, self._name)
+                raise AccessError(_('The requested operation cannot be completed due to security restrictions. Please contact your system administrator.\n\n(Document type: %s, Operation: %s)') % \
+                                  (self._description, operation))
+
+        # If we get here, the invalid records are not in the database.
+        if operation in ('read', 'unlink'):
+            # No need to warn about deleting an already deleted record.
+            # And no error when reading a record that was deleted, to prevent spurious
+            # errors for non-transactional search/read sequences coming from clients.
+            return
+        _logger.info('Failed operation on deleted record(s): %s, uid: %s, model: %s', operation, self._uid, self._name)
+        raise MissingError(_('Missing document(s)') + ':' + _('One of the documents you are trying to access has been deleted, please try again after refreshing.'))
+
+    def _filter_access_rule(self, operation):
```

Comment:
Do we name it _filter_access_rule**s**(), to better convey that it applies the whole set of rules? `_check_access_rule()` was badly named indeed, but maybe we can gradually improve the API, and choose to be consistent with `_check_access_rights()` instead? ;-)

Reply:
Good idea, done.